File: agritwin-ai/backend/app/schemas/soil.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional

class SoilBase(BaseModel):
    soil_type: str
    moisture: float
    ph: float
    nitrogen: str
    phosphorus: str
    potassium: str
    
    n_val: Optional[float] = None
    p_val: Optional[float] = None
    k_val: Optional[float] = None
    ec: Optional[float] = None
    oc: Optional[float] = None
    s: Optional[float] = None
    zn: Optional[float] = None
    fe: Optional[float] = None
    cu: Optional[float] = None
    mn: Optional[float] = None
    b: Optional[float] = None

class SoilCreate(SoilBase):
    farm_id: int

class SoilUpdate(BaseModel):
    soil_type: Optional[str] = None
    moisture: Optional[float] = None
    ph: Optional[float] = None
    nitrogen: Optional[str] = None
    phosphorus: Optional[str] = None
    potassium: Optional[str] = None
    n_val: Optional[float] = None
    p_val: Optional[float] = None
    k_val: Optional[float] = None
    ec: Optional[float] = None
    oc: Optional[float] = None
    s: Optional[float] = None
    zn: Optional[float] = None
    fe: Optional[float] = None
    cu: Optional[float] = None
    mn: Optional[float] = None
    b: Optional[float] = None

class SoilResponse(SoilBase):
    id: int
    farm_id: int
    created_at: datetime

    class Config:
        from_attributes = True

class SoilFertilityRequest(BaseModel):
    N: float
    P: float
    K: float
    ph: float
    ec: float
    oc: float
    S: float
    zn: float
    fe: float
    cu: float
    Mn: float
    B: float

class SoilFertilityResponse(BaseModel):
    prediction: int
    fertility: str
    health_score: int
    status: str

File: services/soil.ts
import api from './api';

export interface SoilData {
  soil_type: string;
  moisture: number;
  ph: number;
  nitrogen: string;
  phosphorus: string;
  potassium: string;
  n_val?: number;
  p_val?: number;
  k_val?: number;
  ec?: number;
  oc?: number;
  s?: number;
  zn?: number;
  fe?: number;
  cu?: number;
  mn?: number;
  b?: number;
  created_at?: string;
}

export interface SoilResponse extends SoilData {
  id: number;
  farm_id: number;
  created_at: string;
}

export interface SoilFertilityResponse {
  prediction: number;
  fertility: string;
  health_score: number;
  status: string;
  model_used?: boolean | string;
  message?: string;
  missing_features?: string[];
}

export async function createSoil(
  data: SoilData & { farm_id: number }
): Promise<SoilResponse> {
  const response = await api.post('/soil/', data);
  return response.data;
}

export async function getSoilByFarm(
  farmId: number | string
): Promise<SoilResponse> {
  const response = await api.get(`/soil/${farmId}`);
  return response.data;
}

export async function updateSoil(
  farmId: number | string,
  data: Partial<SoilData>
): Promise<SoilResponse> {
  const response = await api.put(`/soil/${farmId}`, data);
  return response.data;
}

export async function getSoilFertilityByFarm(
  farmId: number | string
): Promise<SoilFertilityResponse> {
  const response = await api.get(`/soil/${farmId}/fertility`);
  return response.data;
}
